### PR TITLE
Adds support for inequalities

### PR DESCRIPTION
**Pull request by ValentinKaisermayer**
Depends on https://github.com/JuliaSymbolics/Symbolics.jl/pull/717

https://github.com/SciML/ModelingToolkit.jl/issues/1786

### CODE REVIEW

**Comment by ValentinKaisermayer** [comment]
Tests will fail

File: src/systems/optimization/optimizationsystem.jl [hunk 1]
```diff
@@ -247,15 +256,21 @@ function DiffEqBase.OptimizationProblem{iip}(sys::OptimizationSystem, u0map,
         hess_prototype = nothing
     end
 
+    lb = first.(getbounds.(dvs)) # TODO: handle case where bounds are parameters
+    ub = last.(getbounds.(dvs))
+    BT = promote_type(eltype(lb), eltype(ub))
+
     if length(cstr) > 0
-        @named cons_sys = NonlinearSystem(cstr, dvs, ps)
+        @named cons_sys = ConstraintsSystem(cstr, dvs, ps)
```
Comment (ValentinKaisermayer) [inline src/systems/optimization/optimizationsystem.jl:264]: Would be nice to have a simplify step here that reduces the system of equalities and inequalities to a minimal form.

File: test/optimizationsystem.jl [hunk 2]
```diff
@@ -52,41 +52,57 @@ sol = solve(prob, BFGS(initial_stepnorm = 0.0001), allow_f_increases = true)
...
-@test sol.minimum < 1.0
-prob.f.cons(out, sol.minimizer, [1.0, 1.0])
-@test out ≈ [1.0]
+@testset "inequality constraint" begin
+    @variables x y
+    @parameters a b
+    loss = (a - x)^2 + b * (y - x^2)^2
+    cons = [
+        -1 ≲ x^2 + y^2,
+        x^2 + y^2 ≲ 500,
```
Comment (ValentinKaisermayer) [inline test/optimizationsystem.jl:61]: A nicer syntax can come later: `-1 ≲ x^2 + y^2 ≲ 500` 😄

**Comment by YingboMa** [comment]
Try to use a different solver in the tests

**Comment by ValentinKaisermayer** [comment]
Should be good to go now.

File: src/systems/optimization/optimizationsystem.jl [hunk 3]
```diff
@@ -207,29 +209,50 @@ function DiffEqBase.OptimizationProblem{iip}(sys::OptimizationSystem, u0map,
                                              linenumbers = true, parallel = SerialForm(),
                                              use_union = false,
                                              kwargs...) where {iip}
+    if !(isnothing(lb) && isnothing(ub))
+        Base.depwarn("`lb` and `ub` are deprecated. Use the `bounds` metadata for the variables instead.",
```
Comment (Vaibhavdixit02) [inline src/systems/optimization/optimizationsystem.jl:213]: Let's not deprecate, just remove from docs. I don't want to create a mismatch for the constructor, it would just cause confusions that's not worth handling

Comment (Vaibhavdixit02) [reply src/systems/optimization/optimizationsystem.jl:213]: Both bounds and constraints

Comment (ValentinKaisermayer) [reply src/systems/optimization/optimizationsystem.jl:213]: But what should happen if a user supplies values for these?

File: src/systems/optimization/optimizationsystem.jl [hunk 4]
```diff
@@ -207,29 +209,50 @@ function DiffEqBase.OptimizationProblem{iip}(sys::OptimizationSystem, u0map,
...
+    if haskey(kwargs, :lcons) || haskey(kwargs, :ucons)
+        Base.depwarn("`lcons` and `ucons` are deprecated. Specify constraints directly instead.",
+                     :OptimizationProblem, force = true)
+    end
+
     dvs = states(sys)
     ps = parameters(sys)
     cstr = constraints(sys)
 
+    lb = first.(getbounds.(dvs))
```
Comment (Vaibhavdixit02) [inline src/systems/optimization/optimizationsystem.jl:226]:
```suggestion
    lb = lb === nothing ? first(getbounds.(dvs)) : lb
```

File: src/systems/optimization/optimizationsystem.jl [hunk 5]
```diff
@@ -207,29 +209,50 @@ function DiffEqBase.OptimizationProblem{iip}(sys::OptimizationSystem, u0map,
...
 
     u0 = varmap_to_vars(u0map, dvs; defaults = defs, tofloat = false)
     p = varmap_to_vars(parammap, ps; defaults = defs, tofloat = false, use_union)
-    lb = varmap_to_vars(lb, dvs; check = false, tofloat = false, use_union)
-    ub = varmap_to_vars(ub, dvs; check = false, tofloat = false, use_union)
+    lb = varmap_to_vars(dvs .=> lb, dvs; defaults = defs, tofloat = false, use_union)
+    ub = varmap_to_vars(dvs .=> ub, dvs; defaults = defs, tofloat = false, use_union)
+
+    if all(lb .== -Inf) && all(ub .== Inf)
+        lb = nothing
```
Comment (Vaibhavdixit02) [inline src/systems/optimization/optimizationsystem.jl:242]:
```suggestion
        lb = lb
```

File: src/systems/optimization/optimizationsystem.jl [hunk 6]
```diff
@@ -207,29 +209,50 @@ function DiffEqBase.OptimizationProblem{iip}(sys::OptimizationSystem, u0map,
...
+        ub = nothing
```
Comment (Vaibhavdixit02) [inline src/systems/optimization/optimizationsystem.jl:243]:
```suggestion
        ub = ub
```

File: src/systems/optimization/optimizationsystem.jl [hunk 7]
```diff
@@ -409,7 +457,10 @@ function OptimizationProblemExpr{iip}(sys::OptimizationSystem, u0map,
             hess = $_hess
             lb = $lb
             ub = $ub
-            cons = $cons
+            int = $int
+            cons = $cons[1]
+            lbcons = $lbcons
```
Comment (Vaibhavdixit02) [inline src/systems/optimization/optimizationsystem.jl:462]: This should be `lcons` i think?

File: src/systems/optimization/optimizationsystem.jl (hunk 5 above)
Comment (ValentinKaisermayer) [reply src/systems/optimization/optimizationsystem.jl:242]: Actually no, I think. Otherwise some solvers that do not support box constraints will complain.

File: src/systems/optimization/optimizationsystem.jl (hunk 7 above)
Comment (ValentinKaisermayer) [reply src/systems/optimization/optimizationsystem.jl:462]: Oh, yes. My bad.

File: src/systems/optimization/optimizationsystem.jl (hunk 5 above)
Comment (Vaibhavdixit02) [reply src/systems/optimization/optimizationsystem.jl:242]: It would still be nothing right, since the default for the argument is `nothing`?

Comment (Vaibhavdixit02) [reply src/systems/optimization/optimizationsystem.jl:242]: A combination of https://github.com/SciML/ModelingToolkit.jl/pull/1799#discussion_r1010953949 should allow this to allow the user to pass in bounds through the kwarg as well as use the symbolic interface for it. It preserves backward compatibility and keeps the constructor consistent.

Comment (ValentinKaisermayer) [reply src/systems/optimization/optimizationsystem.jl:242]: It would be possible for the box constraints but I think the effort for the constraints is not worth it. It would be very hard to keep in sync. For example if a user specifies `ucons[i] != lcons[i]` for some `i` that corresponds to an equality constraint, we either have to forbid it or we would have to convert it to two inequalities. 

BTW what do you mean with consistent? With what?

Comment (ValentinKaisermayer) [reply src/systems/optimization/optimizationsystem.jl:242]: I think the comfort of writing the constraints directly, like in any good algebraic modelling language like JuMP or GAMS, is worth the depreciation and breaking change.

File: src/systems/optimization/optimizationsystem.jl (hunk 6 above)
Comment (ValentinKaisermayer) [reply src/systems/optimization/optimizationsystem.jl:243]: Unfortunately this does not work

File: src/systems/optimization/optimizationsystem.jl (hunk 3 above)
Comment (YingboMa) [reply src/systems/optimization/optimizationsystem.jl:213]: We should avoid unnecessary breaking changes.

**Review by YingboMa — COMMENTED** [review]
The doc failure looks real. Could you merge master back to your PR?

File: src/systems/optimization/optimizationsystem.jl [hunk 8]
```diff
@@ -354,8 +365,41 @@ function OptimizationProblemExpr{iip}(sys::OptimizationSystem, u0map,
...
+    defs = defaults(sys)
+    defs = mergedefaults(defs, parammap, ps)
+    defs = mergedefaults(defs, u0map, dvs)
+
+    u0 = varmap_to_vars(u0map, dvs; defaults = defs, tofloat = false)
+    p = varmap_to_vars(parammap, ps; defaults = defs, tofloat = false, use_union)
+    lb = varmap_to_vars(dvs .=> lb, dvs; defaults = defs, tofloat = false, use_union)
+    ub = varmap_to_vars(dvs .=> ub, dvs; defaults = defs, tofloat = false, use_union)
+
+    if !isnothing(lb) && all(lb .== -Inf) && !isnothing(ub) && all(ub .== Inf)
```
Comment (Vaibhavdixit02) [inline src/systems/optimization/optimizationsystem.jl:398]: Do it separately for `lb ` and `ub`

Comment (Vaibhavdixit02) [reply src/systems/optimization/optimizationsystem.jl:398]:
```suggestion
    if !isnothing(lb) && all(isinf, lb)
    
```

Comment (Vaibhavdixit02) [reply src/systems/optimization/optimizationsystem.jl:398]: I changed `!isnothing` to `isnothing`

File: src/systems/optimization/optimizationsystem.jl [hunk 9]
```diff
@@ -354,8 +365,41 @@ function OptimizationProblemExpr{iip}(sys::OptimizationSystem, u0map,
                                       linenumbers = false, parallel = SerialForm(),
                                       use_union = false,
                                       kwargs...) where {iip}
+    if haskey(kwargs, :lcons) || haskey(kwargs, :ucons)
```
Comment (Vaibhavdixit02) [inline src/systems/optimization/optimizationsystem.jl:368]: I think we can still allow it @ValentinKaisermayer, the only reason I am being pedantic about these is that I want to avoid divergences between the symbolic interface here and the regular one. That creates a maintenance burden. It would just require the same kind of check as https://github.com/SciML/ModelingToolkit.jl/pull/1799/files#diff-b5294724a3c67d3b2d440b7dba97870040e1a699da417e7b3ac519f91ed61c22R398 after https://github.com/SciML/ModelingToolkit.jl/pull/1799/files#diff-b5294724a3c67d3b2d440b7dba97870040e1a699da417e7b3ac519f91ed61c22R438

Comment (ValentinKaisermayer) [reply src/systems/optimization/optimizationsystem.jl:370]: Somehow the links do not work as intendent - at least for me.

**Comment by Vaibhavdixit02** [comment]
Looks like you need to run the formatter

File: docs/src/systems/OptimizationSystem.md [hunk 10]
```diff
@@ -8,9 +8,10 @@ OptimizationSystem
 
 ## Composition and Accessor Functions
 
-- `get_eqs(sys)` or `equations(sys)`: The equation to be minimized.
+- `get_op(sys)` or `objective(sys)`: The objective to be minimized.
```
Comment (ChrisRackauckas) [inline docs/src/systems/OptimizationSystem.md:11]: What the heck is this for? It ruins the uniformity and then bloats the namespace downstream. It serves no purpose other than to make things less consistent and break downstream codes. Let's revert this change. Just call the equations equations. In fact, there can be more than one, so it's not even the objective, so it's not even a correct description!
